[test] statefulset controller: fix requests tracker concurrency

atiratree opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind bug
/kind flake

#### What this PR does / why we need it:

fixes a flake when using Parallel PodManagementPolicy (burst) in StatefulSets

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #124526

#### Special notes for your reviewer:

tested with the stress tool

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

atiratree committed:
fix requests tracker concurrency

atiratree commented:
/triage accepted
/priority important-longterm

ingvagabund commented:
Can you more elaborate on why the current implementation does not work as expected? I.e. what's the underlying cause that makes the test flake? And what has changed and how it removes the flakiness?

atiratree commented:
We set ecpected failure here after 2 requests: 

https://github.com/kubernetes/kubernetes/blob/0590bb1ac495ae8af2a573f879408e48800da2c5/pkg/controller/statefulset/stateful_set_control_test.go#L523

Then we scale up the replica set in parallel (Burst).

https://github.com/kubernetes/kubernetes/blob/0590bb1ac495ae8af2a573f879408e48800da2c5/pkg/controller/statefulset/stateful_set_control_test.go#L525

Then the pod tracker is called. 

https://github.com/kubernetes/kubernetes/blob/0590bb1ac495ae8af2a573f879408e48800da2c5/pkg/controller/statefulset/stateful_set_control_test.go#L2516

Each tracker method is thread safe but they are not thread safe together. This is the reason behind the flake.

So we fix that by locking the increment and error checking with a single lock.

---

Also we have the parallel requests checking which is independent and used in TestParallelScale test.

https://github.com/kubernetes/kubernetes/blob/0590bb1ac495ae8af2a573f879408e48800da2c5/pkg/controller/statefulset/stateful_set_control_test.go#L2946

This was guarded by a lock before and maxParallel was always equal to the number of requests. So it didn't test the parallelness correctly and had to be fixed as well.

ingvagabund commented on pkg/controller/statefulset/stateful_set_control_test.go:
```diff
@@ -2423,51 +2423,48 @@ type requestTracker struct {
+		// reset and pass the error
+		defer func() {
+			rt.err = nil
+			rt.after = 0
```
`reset()` originally set rt.delay = 0. Why not do the same here for `rt.parallelRequestDelay`?

atiratree replied:
As mentioned in https://github.com/kubernetes/kubernetes/pull/124624#issuecomment-2098186385 
this reset was never used before, because errors and parallel requests have not been tested at the same time before.

Logic of these two blocks `trackParallelRequests` and `incWithOptionalError` is independent of each other, so I think it makes sense to keep them separate.

ingvagabund commented on pkg/controller/statefulset/stateful_set_control_test.go:
```diff
@@ -2945,47 +2938,52 @@ func fakeResourceVersion(object interface{}) {
+		desc                        string
+		replicas                    int32
+		desiredReplicas             int32
+		expectedMinParallelRequests int
```
`expectedMinParallelRequests` is a new field? Why the original `1` is not enough?

atiratree replied:
I have moved this refactoring to a new commit: https://github.com/kubernetes/kubernetes/pull/124624/commits/7b725f3bfcc0eb88a75f91f523e7b669213321ef.

This is to make the test more accurate as originally we expected only 2 parallel requests. But if we create a 1000 pods we should expect more, otherwise it would suggest a regression.

ingvagabund commented on pkg/controller/statefulset/stateful_set_control_test.go:
```diff
@@ -3017,8 +3015,8 @@ func parallelScale(t *testing.T, set *apps.StatefulSet, replicas, desiredReplica
 
-	if (diff < -1 || diff > 1) && om.createPodTracker.maxParallel <= 1 {
-		t.Errorf("want max parallel requests > 1, got %v", om.createPodTracker.maxParallel)
+	if om.createPodTracker.maxParallelRequests < expectedMinParallelRequests {
```
Why is `(diff < -1 || diff > 1)` no longer tested?

ingvagabund commented on pkg/controller/statefulset/stateful_set_control_test.go:
```diff
@@ -2423,51 +2423,48 @@ type requestTracker struct {
 
-func (rt *requestTracker) reset() {
-	rt.parallelLock.Lock()
-	rt.parallel = 0
```
Turning into `rt.parallelRequests`. So from now on `rt.parallelRequests` is never reset. So originally `rt.parallel` was counting the number of requests in parallel until an error was ready. Followed by resting it. Why is it no longer needed to reset it?

atiratree replied:
https://github.com/kubernetes/kubernetes/pull/124624#discussion_r1592923857

ingvagabund commented:
Worth mentioning:
- with `trackParallelRequests` locking the parallel requests/routines are serialized which makes the testing less real (some of the state space is not explored).
- removing reset of both delay and parallel fields is ok since they execution path does not generate the error so the reset code is actually never executed -> safe to remove the code resetting both fields

atiratree committed:
improve TestParallelScale test

we can test for larger number of parallel requests than 2 in
statefulsets with hundreds of replicas

atiratree commented:
> with trackParallelRequests locking the parallel requests/routines are serialized which makes the testing less real (some of the state space is not explored).

I have solved this by requiring the enablement for the tracking

https://github.com/kubernetes/kubernetes/pull/124624/files#diff-49263a54d6a7753728e489881537462578ddc2b9b09404ee335c73e27e178ddcR2436-R2438

so it should not affect the other tests now, apart from the `TestParallelScale`

---

I have also added a check for the expectedMinParallelRequests values.

https://github.com/kubernetes/kubernetes/pull/124624/files#diff-49263a54d6a7753728e489881537462578ddc2b9b09404ee335c73e27e178ddcR2990-R2994

ingvagabund commented:
/lgtm
